File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/highriskvenuesupload/model/HighRiskVenue.java
package uk.nhs.nhsx.highriskvenuesupload.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Objects;
@JsonInclude(JsonInclude.Include.NON_EMPTY)
public class HighRiskVenue {
    public final String id;
    public final RiskyWindow riskyWindow;
    public final String messageType;
    public final String optionalParameter;

    public HighRiskVenue(String id, RiskyWindow riskyWindow) {
        this(id, riskyWindow, "M1", null);
    }
    @JsonCreator
    public HighRiskVenue(String id, RiskyWindow riskyWindow, String messageType,  String optionalParameter) {
        this.id = id;
        this.riskyWindow = riskyWindow;
        this.messageType = messageType;
        this.optionalParameter = optionalParameter;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        HighRiskVenue that = (HighRiskVenue) o;
        return Objects.equals(id, that.id) &&
            Objects.equals(riskyWindow, that.riskyWindow) &&
            Objects.equals(messageType, that.messageType) &&
            Objects.equals(optionalParameter, that.optionalParameter);
    }

    @Override
    public String toString() {
        return "HighRiskVenue{" +
            "id='" + id + '\'' +
            ", riskyWindow=" + riskyWindow +
            ", messageType='" + messageType + '\'' +
            ", optionalParameter='" + optionalParameter + '\'' +
            '}';
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, riskyWindow, messageType, optionalParameter);
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/highriskvenuesupload/model/HighRiskVenue.kt
package uk.nhs.nhsx.highriskvenuesupload.model

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL
import dev.forkhandles.values.NonBlankStringValueFactory
import dev.forkhandles.values.StringValue


class OptionalHighRiskVenueParam(val type: String): StringValue(type) {
    companion object : NonBlankStringValueFactory<OptionalHighRiskVenueParam>(::OptionalHighRiskVenueParam)
}

@JsonInclude(NON_NULL)
data class HighRiskVenue(
    val id: VenueId,
    val riskyWindow: RiskyWindow,
    val messageType: MessageType,
    val optionalParameter: OptionalHighRiskVenueParam? = null
) {
    init {
        if (!MESSAGE_TYPES_WITH_OPTIONAL_PARAMETER.contains(messageType) && !(optionalParameter?.value.isNullOrEmpty())) {
            throw IllegalArgumentException("validation error: Message type $messageType does not support optional parameter")
        }

        if (MESSAGE_TYPES_WITH_OPTIONAL_PARAMETER.contains(messageType) && (optionalParameter?.value.isNullOrEmpty())) {
            throw  IllegalArgumentException("validation error: Message type $messageType must include an optional parameter")
        }
    }

    companion object {
        private val MESSAGE_TYPES_WITH_OPTIONAL_PARAMETER = emptyList<MessageType>()
    }
}
